From: Dave Reisner <dreisner@archlinux.org>
Date: Sat, 8 Feb 2014 12:54:58 -0500
Subject: cryptsetup-generator: auto add deps for device as password

If the password is a device file, we can add Requires/After dependencies
on the device rather than requiring the user to do so.

(cherry picked from commit 66a5dbdf282435403f947c2caadd04bb0cdec752)
---
 src/cryptsetup/cryptsetup-generator.c | 20 +++++++++++++++-----
 1 file changed, 15 insertions(+), 5 deletions(-)

diff --git a/src/cryptsetup/cryptsetup-generator.c b/src/cryptsetup/cryptsetup-generator.c
index e1798a3..00891f1 100644
--- a/src/cryptsetup/cryptsetup-generator.c
+++ b/src/cryptsetup/cryptsetup-generator.c
@@ -131,11 +131,21 @@ static int create_disk(
                     streq(password, "/dev/random") ||
                     streq(password, "/dev/hw_random"))
                         fputs("After=systemd-random-seed.service\n", f);
-                else if (!streq(password, "-") &&
-                         !streq(password, "none"))
-                        fprintf(f,
-                                "RequiresMountsFor=%s\n",
-                                password);
+
+                else if (!streq(password, "-") && !streq(password, "none")) {
+                        _cleanup_free_ char *uu = fstab_node_to_udev_node(password);
+                        if (uu == NULL)
+                                return log_oom();
+
+                        if (is_device_path(uu)) {
+                                _cleanup_free_ char *dd = unit_name_from_path(uu, ".device");
+                                if (dd == NULL)
+                                        return log_oom();
+
+                                fprintf(f, "After=%1$s\nRequires=%1$s\n", dd);
+                        } else
+                                fprintf(f, "RequiresMountsFor=%s\n", password);
+                }
         }
 
         if (is_device_path(u))
